File: Array/medium/34.cpp
// Runtime: 8 ms, faster than 85.56% of C++ online submissions for Find First and Last Position of Element in Sorted Array.
// Memory Usage: 9.8 MB, less than 100.00% of C++ online submissions for Find First and Last Position of Element in Sorted Array.

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if (nums.empty())
            return vector<int>({-1, -1});
        
        // find one of it by standard binary search
        int low = 0;
        int high = nums.size() - 1;
        int mid;
        while (low <= high) {
            mid = (low + high) / 2;
            if (nums[mid] == target) break;
            else if (nums[mid] > target) high = mid - 1;
            else low = mid + 1;
        }
        int targetSpot = mid;
        
        cout << mid << endl;
        
        // not found
        if (nums[targetSpot] != target)
            return vector<int>({-1, -1});
        
        // find left
        low = 0;
        while (low < mid) {
            int leftMid = (low + mid) / 2;
            
            if (nums[leftMid] == target) mid = leftMid;
            else low = leftMid + 1;
        }
        
        cout << low << endl;
        
        // find right
        mid = targetSpot; // reset
        high = nums.size() - 1;
        while (mid < high) {
            int rightMid = (mid + high) / 2 + (mid + high) % 2; // flooring
            
            if (nums[rightMid] == target) mid = rightMid;
            else high = rightMid - 1;
        }
        cout << high << endl;
        
        return vector<int>({low, high});
    }
};

File: BinarySearch/medium/34.py
"""
Naive:
binary search to find target in nums, cost O(logn)

then linear reach out the left and right end of target, cost O(n) at worst case


Improvement:
how about find target+0.5 and target-0.5?


edge cases:
1. nums:[1,1,1], t: 1

"""


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearch(t):
            low = 0
            high = len(nums) - 1

            while low <= high:
                mid = (low + high) // 2
                if nums[mid] < t:
                    low = mid + 1
                elif nums[mid] > t:
                    high = mid - 1

            return high

        leftTarget = target - 0.5
        rightTarget = target + 0.5

        left = binarySearch(leftTarget) + 1
        right = binarySearch(rightTarget)
        return [left, right] if left <= right else [-1, -1]
